### PR TITLE
Create a scatter plot for the emission profile and transition risk, with financial data

### LINKED ISSUE #85: New plot: Create new scatter plot
@Tilmon and I came up with a new scatter plot. The scatter plot should look like this: 

Image

Please note that this work depends on the transition risk score. That's why it will be something for next week.

### PR DESCRIPTION
**Pull request by lindadelacombaz**
Closes #85 and #102.

Dear @maurolepore, 
Here is the scatter plot I have created for the financial data.

The financial data changed, but it breaks my code for the emission bar plots with financial data (hence the empty image in the README file). I will explore this in another PR.
Maybe it is also time for me to create more unit tests so I can prevent this to happen.

Thank you ! :))

----

TODO

- [x] [Link related issue/PR]([url](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). 
- [x] Describe the goal of the PR. Avoid details that are clear in the diff.
- [x] Mark the PR as draft.
- [x] [Include a unit test](https://code-review.tidyverse.org/reviewer/aspects.html#sec-tests).
- [x] Review your own PR in "Files changed".
- [x] Ensure the PR branch is updated.
- [x] Ensure the checks pass.
- [x] Change the status from draft to ready.
- [x] Polish the PR title and description.
- [x] Assign a reviewer.

EXCEPTIONS

- [ ] Slide here any item that you intentionally choose to not do.

### CODE REVIEW

File: R/scatter_plot_financial.R [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+  data <- process_data(data, benchmarks_arg, scenario_arg, year_arg)
+
+  emission_rank <- calculate_rank(data, mode_var, "profile_ranking")
+  tr_score <- calculate_rank(data, mode_var, "transition_risk_score")
+
+  ggplot(data, aes(x = .data$amount_total, color = .data$bank_id)) +
+    geom_point(aes(y = emission_rank, shape = "Emission Rank")) +
+    geom_point(aes(y = tr_score, shape = "TR Score")) +
+    scale_shape_manual(name = "Legend", values = c("Emission Rank" = 17, "TR Score" = 18)) +
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:77]: I see very specific values like "Emission Rank" and `17`. I'm not sure what they do but it seems brittle. Would the code resist a change, say from "Emission Rank" to "Emissions Rank" or "Emission rank"? If not, what's the most maintainable way to program this?

File: R/scatter_plot_financial.R [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+#' @param col A character vector.
+#'
+#' @return A numerical value.
+#' @noRd
+calculate_rank <- function(data, mode, col) {
+  if (mode == "equal_weight_finance") {
+    rank <- mean(data[[col]])
+  } else if (mode %in% c("worst_case_finance", "best_case_finance")) {
+    df <- data[data[[mode]] != 0, ]
+    rank <- mean(data[[col]])
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:98]: Do we need to consider missing values? i.e. `mean(..., na.rm = TRUE)` or some other way to handle `NA`s? Make sure to test edge cases.

Comment (maurolepore) [inline R/scatter_plot_financial.R:98]: Where is `df` used? I think it's defined but never used.

File: R/scatter_plot_financial.R [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+#' Calculate Rank
+#'
+#' @param data A data frame.
+#' @param mode A character vector.
+#' @param col A character vector.
+#'
+#' @return A numerical value.
+#' @noRd
+calculate_rank <- function(data, mode, col) {
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:93]: When using `if` with strings you may consider `switch()`, e.g.

```r
rank <- switch(mode,
  "equal_weight_finance" = mean(data[[col]]),
  "worst_case_finance" = ,
  "best_case_finance" = {
    # FIXME: `df` is defined but never used
    df <- data[data[[mode]] != 0, ]
    mean(data[[col]])
  }
)
```

https://speakerdeck.com/jennybc/code-smells-and-feels?slide=63

File: R/scatter_plot_financial.R [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+#' Process data
+#'
+#' @param data A data frame.
+#' @param benchmarks_arg A character vector.
+#' @param scenario_arg A character vector.
+#' @param year_arg A numerical value.
+#'
+#' @return A data frame.
+#' @noRd
+process_data <- function(data, benchmarks_arg, scenario_arg, year_arg) {
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:112]: Here again we may remove the suffix `_arg`. I think it introduces a new name for a concept that is already well captured by "bemchmark", "scenario", and "year", and also it's somewhat redundant (with the same logic the `data` argument should become `data_arg`, or `data_df`?).

File: R/scatter_plot_financial.R [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      .data$scenario == scenario_arg,
+      .data$year == year_arg
+    )
+
+  data <- data |>
+    group_by(.data$tilt_sector) |>
+    mutate(percent = mean(.data$reduction_targets)) |>
+    mutate(
+      percent = round(.data$percent * 100, 4),
+      title = glue::glue("{unique(.data$tilt_sector)}: {unique(.data$percent)}% SERT")
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:125]: :-)

File: R/scatter_plot_financial.R [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  data <- data |>
+    filter(
+      .data$benchmark %in% benchmarks_arg,
+      .data$scenario == scenario_arg,
+      .data$year == year_arg
+    )
+
+  data <- data |>
+    group_by(.data$tilt_sector) |>
+    mutate(percent = mean(.data$reduction_targets)) |>
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:122]: Every time I see `mean()` I remember the `na.rm` argument, then I remember I should think how to handle missing values, then I remember to write tests for edge cases.

File: R/scatter_plot_financial.R [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+#' @param scenario A character vector specifying the scenario to be plot. It can
+#' either be "IPR" or "WEO".
+#' @param year A numerical specifying the year of the scenario to be plot. It
+#' can be either 2030 or 2050.
+#'
+#' @return A [ggplot] object.
+#' @export
+#'
+#' @examples
+#' scatter_plot_financial(financial, "all", "equal_weight", "IPR", 2030)
```
Comment (maurolepore) [inline R/scatter_plot_financial.R:22]: In general it's fine to omit the name of the first argument (easiest to remember) but the second and others should be named explicitely. Pretty sure I read that advice somewhere in https://design.tidyverse.org/ or https://style.tidyverse.org/

```r
scatter_plot_financial(financial, benchmarks = "all", mode = "equal_weight", scenario = "IPR", year = 2030)
```

As I type this I realize that the user should choose only 1 benchmark so the plural "benchmark**s**" seems missleading.

**Review by maurolepore — APPROVED** [review]
LGTM but see my  comments. 

And yeah, more tests would be nice :-)

As you extract more and more code into preparaiton functions it seems possible to think of them as an internal API for which you can write all sorts of tests (you may even extract them into their own package). That internal API doesn't deal with plots at all, so it's much easier to test.

File: R/scatter_plot_financial.R [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                                     "unit_isic_4digit",
+                                     "unit_tilt_sector"
+                                   ),
+                                   mode = c(
+                                     "equal_weight",
+                                     "worst_case",
+                                     "best_case"
+                                   ),
+                                   scenario = c("IPR", "WEO"),
+                                   year = c(2030, 2050)) {
```
[the remark this answers is not in the document]
Comment (lindadelacombaz) [reply R/scatter_plot_financial.R:38]: I am unsure of the difference of using @noRd and @keywords internal. For now I always used @noRd for my helpers functions. 
From what I have read, @keywords internal creates a [manual page but is not present in the function index](https://blog.r-hub.io/2019/12/12/internal-functions/) @noRd simply does not create manual pages. 
I am not sure to understand when to use which ?

File: R/scatter_plot_financial.R (hunk 7 above)
Comment (lindadelacombaz) [reply R/scatter_plot_financial.R:22]: @maurolepore Thanks ! Good point, @AnneSchoenauer asked that the user could choose multiple benchmarks for some plots (for example the bar plot.) I will check with her if that is the case for the scatter plot here.
